flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricQueryService.java

```diff
@@ -0,0 +1,180 @@
+					@Override
+					public String filterCharacters(String input) {
+						return input.replaceAll("[ :.,]", "_");
+					}
+				});
+
+				if (metric instanceof Counter) {
+					counters.put((Counter) metric, name);
+				} else if (metric instanceof Gauge) {
+					gauges.put((Gauge<?>) metric, name);
+				} else if (metric instanceof Histogram) {
+					histograms.put((Histogram) metric, name);
+				}
+			} else if (message instanceof Metric) { // remove metric
+				toRemove.add((Metric) message);
+			} else if (message instanceof ActorRef) { // create dump
+				Object[] dump = createDump();
+				((ActorRef) message).tell(dump, getSelf());
+				removeMetrics();
+			} else {
+				LOG.warn("MetricQueryServiceActor received an invalid message. " + message.toString());
+			}
+		} catch (Exception e) {
+			LOG.warn("An exception occurred while processing a message.", e);
+		}
+	}
+
+	private Object[] createDump() {
+		Object[] metrics = new Object[(this.counters.size() + this.gauges.size()) * 3 + this.histograms.size() * 13];
+
+		int index = 0;
+
+		// dump counters
+		for (Map.Entry<Counter, String> counter : this.counters.entrySet()) {
+			metrics[index++] = CATEGORY_COUNTER;
+			metrics[index++] = counter.getValue();
+			metrics[index++] = counter.getKey().getCount();
+		}
+		// dump gauges
+		for (Map.Entry<Gauge<?>, String> gauge : this.gauges.entrySet()) {
+			metrics[index++] = CATEGORY_GAUGE;
+			metrics[index++] = gauge.getValue();
+			metrics[index++] = gauge.getKey().getValue().toString();
+		}
+		// dump histograms
+		for (Map.Entry<Histogram, String> histogram : this.histograms.entrySet()) {
+			HistogramStatistics stat = histogram.getKey().getStatistics();
+
+			metrics[index++] = CATEGORY_HISTOGRAM;
+			metrics[index++] = histogram.getValue();
+			metrics[index++] = stat.getMin();
+			metrics[index++] = stat.getMax();
+			metrics[index++] = stat.getMean();
+			metrics[index++] = stat.getQuantile(0.5);
+			metrics[index++] = stat.getStdDev();
+			metrics[index++] = stat.getQuantile(0.75);
+			metrics[index++] = stat.getQuantile(0.9);
+			metrics[index++] = stat.getQuantile(0.95);
+			metrics[index++] = stat.getQuantile(0.98);
+			metrics[index++] = stat.getQuantile(0.99);
+			metrics[index++] = stat.getQuantile(0.999);
+		}
+		return metrics;
+	}
+
+	private void removeMetrics() {
+		for (Metric metric : toRemove) {
+			if (metric instanceof Counter) {
+				this.counters.remove(metric);
+			} else if (metric instanceof Gauge) {
+				this.gauges.remove(metric);
+			} else if (metric instanceof Histogram) {
+				this.histograms.remove(metric);
+			}
+		}
+	}
+
+	/**
+	 * Starts the MetricQueryService actor in the given actor system.
+	 *
+	 * @param actorSystem The actor system running the MetricQueryService
+	 * @return actor reference to the MetricQueryService
+	 */
+	public static ActorRef startMetricQueryService(ActorSystem actorSystem) {
+		return actorSystem.actorOf(Props.create(MetricQueryService.class));
```

Comment:
`actorSystem.actorOf(Props.create(MetricQueryService.class), "foobar")` creates an actor with the name "foobar". If your actor system is the Flink actor system, then it should be reachable via `akka.tcp://flink@xxx.yyy.zzz:port/foobar` if I'm not mistaken. But you would need an actorsystem reference in the web server to do the actor resolution.